module/dkg/broker.go

```diff
@@ -312,6 +308,18 @@ func (b *Broker) updateContractClient(clientIndex int, init bool) (int, module.D
 	return clientIndex, b.dkgContractClients[clientIndex]
 }
 
+// getInitialContractClient will return the last successful contract client or the initial
+func (b *Broker) getInitialContractClient() (int, module.DKGContractClient) {
+	return b.lastSuccessfulClientIndex, b.dkgContractClients[b.lastSuccessfulClientIndex]
```

Comment:
⚠️  Reads of the shared variable `lastSuccessfulClientIndex ` need to be synchronized as well. 

Currently the code will prevent concurrent _writes_ to `lastSuccessfulClientIndex`, but will allow concurrently one write and any number of reads. 

In general under the Go memory model, this results in undefined behaviour, from the [Go Memory Model](https://golang.org/ref/mem#tmp_10):
> Programs that modify data being simultaneously accessed by multiple goroutines must serialize such access.

Reply:
+1 on this, strictly speaking this might not a problem. Since the write is protected by `Mutex` we will observe changes in other goroutines. I haven't found proof for that behavior in memory model, but go core devs use this [trick](https://github.com/golang/go/blob/master/src/runtime/chan.go#L138) and expect such behavior.

Just to be clear, I am not saying that we should do this, we should just place an explicit lock and be sure that it's safe and obvious, this off-topic is more for knowledge sharing of interesting fact.

Reply:
https://github.com/onflow/flow-go/pull/1556/commits/6f9ab2af46584eb4c8fbea55778896dcbe0e2d9c